File: DashWallet/Sources/Models/DWDateFormatter.m
//
//  Created by Andrew Podkovyrin
//  Copyright © 2020 Dash Core Group. All rights reserved.
//
//  Licensed under the MIT License (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://opensource.org/licenses/MIT
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "DWDateFormatter.h"

NS_ASSUME_NONNULL_BEGIN

@interface DWDateFormatter ()

@property (readonly, nonatomic, strong) NSDateFormatter *shortDateFormatter;
@property (readonly, nonatomic, strong) NSDateFormatter *longDateFormatter;
@property (readonly, nonatomic, strong) NSDateFormatter *iso8601DateFormatter;

@end

NS_ASSUME_NONNULL_END

@implementation DWDateFormatter

+ (instancetype)sharedInstance {
    static DWDateFormatter *_sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _sharedInstance = [[self alloc] init];
    });
    return _sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        NSLocale *locale = [NSLocale currentLocale];
        _shortDateFormatter = [[NSDateFormatter alloc] init];
        _shortDateFormatter.dateFormat = [NSDateFormatter dateFormatFromTemplate:@"MMMdjmma"
                                                                         options:0
                                                                          locale:locale];
        _longDateFormatter = [[NSDateFormatter alloc] init];
        _longDateFormatter.dateFormat = [NSDateFormatter dateFormatFromTemplate:@"yyyyMMMdjmma"
                                                                        options:0
                                                                         locale:locale];

        _iso8601DateFormatter = [[NSDateFormatter alloc] init];
        NSLocale *enUSPOSIXLocale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
        [_iso8601DateFormatter setLocale:enUSPOSIXLocale];
        [_iso8601DateFormatter setDateFormat:@"yyyy-MM-dd'T'HH:mm:ss'Z'"];
        [_iso8601DateFormatter setCalendar:[NSCalendar calendarWithIdentifier:NSCalendarIdentifierGregorian]];
    }
    return self;
}

- (NSString *)shortStringFromDate:(NSDate *)date {
    NSCalendar *calendar = [NSCalendar currentCalendar];
    NSInteger nowYear = [calendar component:NSCalendarUnitYear fromDate:[NSDate date]];
    NSInteger dateYear = [calendar component:NSCalendarUnitYear fromDate:date];

    NSDateFormatter *desiredFormatter = (nowYear == dateYear) ? self.shortDateFormatter : self.longDateFormatter;
    return [desiredFormatter stringFromDate:date];
}

- (NSString *)longStringFromDate:(NSDate *)date {
    return [self.longDateFormatter stringFromDate:date];
}

- (NSString *)iso8601StringFromDate:(NSDate *)date {
    return [self.iso8601DateFormatter stringFromDate:date];
}

@end

File: DashWallet/Sources/Models/DWDateFormatter.swift
//
//  Created by Andrew Podkovyrin
//  Copyright © 2019 Dash Core Group. All rights reserved.
//
//  Licensed under the MIT License (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://opensource.org/licenses/MIT
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

@objc
class DWDateFormatter: NSObject {
    @objc
    static let sharedInstance = DWDateFormatter()

    private let shortDateFormatter: DateFormatter
    private let longDateFormatter: DateFormatter
    private let iso8601DateFormatter: DateFormatter
    private let timeOnlyFormatter: DateFormatter

    private override init() {
        let locale = Locale.current

        shortDateFormatter = DateFormatter()
        shortDateFormatter.dateFormat = DateFormatter.dateFormat(fromTemplate: "dd MMMM", options: 0, locale: locale)

        longDateFormatter = DateFormatter()
        longDateFormatter.dateFormat = DateFormatter.dateFormat(fromTemplate: "dd MMMM, yyyy", options: 0, locale: locale)

        iso8601DateFormatter = DateFormatter()
        let enUSPOSIXLocale = Locale(identifier: "en_US_POSIX")
        iso8601DateFormatter.locale = enUSPOSIXLocale
        iso8601DateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        iso8601DateFormatter.calendar = Calendar(identifier: .gregorian)
        
        timeOnlyFormatter = DateFormatter()
        timeOnlyFormatter.timeStyle = .short
        timeOnlyFormatter.dateStyle = .none
        timeOnlyFormatter.locale = Locale.current
    }

    func dateOnly(from date: Date) -> String {
        let calendar = Calendar.current
        let nowYear = calendar.component(.year, from: Date())
        let dateYear = calendar.component(.year, from: date)

        let desiredFormatter = (nowYear == dateYear) ? shortDateFormatter : longDateFormatter
        return desiredFormatter.string(from: date)
    }

    func longString(from date: Date) -> String {
        return longDateFormatter.string(from: date)
    }

    func iso8601String(from date: Date) -> String {
        return iso8601DateFormatter.string(from: date)
    }
    
    func timeOnly(from dateTime: Date) -> String {
        return timeOnlyFormatter.string(from: dateTime)
    }
}

@objc extension DWDateFormatter {
    @objc func shortStringFromDate(_ date: Date) -> String {
        return dateOnly(from: date)
    }
}
